File: demos/05-cli-sdk/02-sdk/02-custom-tools/sdk-demos-solution/interactive-agent.ts
import { CopilotClient, defineTool, SessionEvent } from "@github/copilot-sdk";
import { createInterface } from "node:readline";

const getWeather = defineTool("get_weather", {
  description: "Get the current weather for a city",
  skipPermission: true,
  parameters: {
    type: "object",
    properties: {
      city: {
        type: "string",
        description: "The city name",
      },
    },
    required: ["city"],
  },
  handler: async ({ city }: { city: string }) => {
    const conditions = ["sunny", "cloudy", "rainy", "partly cloudy"];
    const temp = Math.floor(Math.random() * 30) + 50;
    const condition = conditions[Math.floor(Math.random() * conditions.length)];
    return { city, temperature: `${temp}°F`, condition };
  },
});

async function main() {
  const client = new CopilotClient();
  const session = await client.createSession({
    model: "gpt-5-mini",
    streaming: true,
    tools: [getWeather],
  });

  session.on((event: SessionEvent) => {
    if (event.type === "assistant.message_delta") {
      process.stdout.write(event.data.deltaContent);
    }
  });

  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: "You: ",
  });

  let closed = false;
  rl.on("close", () => {
    closed = true;
  });
  process.on("SIGINT", () => rl.close());

  console.log("Weather Assistant (type 'exit' to quit)");
  console.log("Try: 'What's the weather in Paris and London?'\n");
  rl.prompt();

  for await (const line of rl) {
    const input = line.trim();
    if (input.toLowerCase() === "exit") {
      break;
    }
    if (input.length > 0) {
      process.stdout.write("Assistant: ");
      await session.sendAndWait({ prompt: input });
      console.log("\n");
    }
    if (closed) {
      break;
    }
    rl.prompt();
  }

  rl.close();
  await client.stop();
}

main().catch((error) => {
  console.error("Error:", error);
  process.exit(1);
});

File: demos/05-cli-sdk/02-sdk/02-custom-tools/custom-tools-solution-py/interactive_agent.py
import asyncio
import random
import sys

from copilot import CopilotClient, define_tool
from copilot.session_events import AssistantMessageDeltaData
from pydantic import BaseModel, Field


class GetWeatherParams(BaseModel):
    city: str = Field(description="The city name")


@define_tool("get_weather", description="Get the current weather for a city", skip_permission=True)
async def get_weather(params: GetWeatherParams) -> dict:
    conditions = ["sunny", "cloudy", "rainy", "partly cloudy"]
    temp = random.randint(50, 79)
    condition = random.choice(conditions)
    return {"city": params.city, "temperature": f"{temp}°F", "condition": condition}


def on_event(event):
    match event.data:
        case AssistantMessageDeltaData() as data:
            print(data.delta_content, end="", flush=True)


async def main():
    client = CopilotClient()
    session = await client.create_session(
        model="gpt-5-mini",
        streaming=True,
        tools=[get_weather],
    )

    session.on(on_event)

    print("Weather Assistant (type 'exit' to quit)")
    print("Try: 'What's the weather in Paris and London?'\n")

    while True:
        print("You: ", end="", flush=True)
        line = await asyncio.to_thread(sys.stdin.readline)
        if not line:
            break
        user_input = line.strip()
        if user_input.lower() == "exit":
            break
        if user_input:
            print("Assistant: ", end="", flush=True)
            await session.send_and_wait(user_input)
            print("\n")

    await client.stop()


asyncio.run(main())
